src/cudamatrix/cu-sparse-matrix.cc

```diff
@@ -42,6 +42,218 @@
+  }
+}
+
+template <typename Real>
+void CuRowSparseMatrix<Real>::SetRandn(BaseFloat zero_prob) {
+  if (num_rows_ == 0) return;
+  // Use the CPU function for the moment, not efficient...
+  SparseMatrix<Real> tmp(num_rows_, num_cols_);
+  tmp.SetRandn(zero_prob);
+  Swap(&tmp);
+}
+
+template <typename Real>
+void CuRowSparseMatrix<Real>::Write(std::ostream &os, bool binary) const {
+  SparseMatrix<Real> tmp;
+  CopyToSmat(&tmp);
+  tmp.Write(os, binary);
+}
+
+template <typename Real>
+void CuRowSparseMatrix<Real>::Read(std::istream &is, bool binary) {
+  SparseMatrix<Real> tmp;
+  tmp.Read(is, binary);
+  this->Swap(&tmp);
+}
+
+template <typename Real>
+template <typename OtherReal>
+void CuRowSparseMatrix<Real>::CopyFromSmat(const SparseMatrix<OtherReal> &smat) {
+  num_rows_ = smat.NumRows();
+  num_cols_ = smat.NumCols();
+  std::vector<MatrixIndexT> cpu_elements_per_row;
+  cpu_elements_per_row.resize(num_rows_);
+  elements_per_row_.Resize(0);
+  MatrixIndexT max_num_elements = 0;
+  for (int32 i = 0; i < num_rows_; ++i) {
+    MatrixIndexT num_elements = (smat.Data() + i)->NumElements();
+    cpu_elements_per_row[i] = num_elements;
+    if (num_elements > max_num_elements)
+      max_num_elements = num_elements;
+  }
+  elements_per_row_.CopyFromVec(cpu_elements_per_row);
+  stride_ = max_num_elements;
+#if HAVE_CUDA == 1
+  if (CuDevice::Instantiate().Enabled()) {
+    if (num_rows_ == 0 || num_cols_ == 0 || stride_ == 0) {
+      data_.Resize(0);
+      return;
+    }
+    RowElement<Real> pad;
+    pad.column = -1;
+    pad.weight = 0.0;
+    std::vector<RowElement<Real> > cpu_elements(num_rows_ * stride_, pad);
+    for (int32 i = 0; i < num_rows_; ++i) {
+      for (int32 j = 0; j < (smat.Data() + i)->NumElements(); ++j) {
+        RowElement<Real>* cpu_element = &(cpu_elements[i * stride_ + j]);
+        cpu_element->column = ((smat.Data() + i)->Data() + j)->first;
+        cpu_element->weight = ((smat.Data() + i)->Data() + j)->second;
+      }
+    }
+    data_.CopyFromVec(cpu_elements);
+  } else
+#endif
+  {
+    this->Mat().CopyFromSmat(smat);
+  }
+}
+template
+void CuRowSparseMatrix<float>::CopyFromSmat(const SparseMatrix<float> &smat);
+template
+void CuRowSparseMatrix<float>::CopyFromSmat(const SparseMatrix<double> &smat);
+template
+void CuRowSparseMatrix<double>::CopyFromSmat(const SparseMatrix<float> &smat);
+template
+void CuRowSparseMatrix<double>::CopyFromSmat(const SparseMatrix<double> &smat);
+
+template <typename Real>
+template <typename OtherReal>
+void CuRowSparseMatrix<Real>::CopyToSmat(SparseMatrix<OtherReal> *smat) const {
+  KALDI_ASSERT(smat != NULL);
+#if HAVE_CUDA == 1
+  if (CuDevice::Instantiate().Enabled()) {
+    std::vector<RowElement<Real> > cpu_data;
+    data_.CopyToVec(&cpu_data);
+    std::vector<std::vector<std::pair<MatrixIndexT, Real> > > pairs(num_rows_);
+    for (int32 i = 0; i < cpu_data.size(); ++i) {
+      pairs[i].push_back(std::make_pair(cpu_data[i].column, cpu_data[i].weight));
```

Comment:
@naxingyu, shouldn't this line be something like pairs[i/stride_] ? Currently index 'i' can get out of bounds of pairs.